src/Chapter3.hs

```diff
@@ -480,6 +524,46 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Inhabitants = One | Two | Three | Four deriving (Show)
+inhabitantsToInt :: Inhabitants -> Int
+inhabitantsToInt One   = 1
+inhabitantsToInt Two   = 2
+inhabitantsToInt Three = 3
+inhabitantsToInt Four  = 4
+
+
+data Castle = Castle String deriving (Show)
+data Wall = Wall deriving (Show)
+data CommunityBuilding = Library | Church deriving (Show)
+data House = House {houseInhabitants :: Inhabitants} deriving (Show)
+
+data CityType
+    = WalledCastleCity Castle Wall CommunityBuilding [House]
+    | CastleCity Castle CommunityBuilding [House]
+    | Town CommunityBuilding [House]
+    deriving (Show)
```

Comment:
I feel like it would be more convenient to work `City` if it was a record type instead of a sum type. You already have `Castle` that stores walls inside. 

First, we need to decide with the castle. Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. But you need to define such type separately (as you did):

```haskell
data Castle = Castle 
    { castleName :: String
    , castleWall :: Bool 
    }
```

And in the city, we will have `Maybe Castle`.

So, instead of having two top-level constructors and duplicating all fields on both of them, you can define `City` as a record with optional castle:

After all that is done, Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- OR you can do it with your type
    -- cityCastle :: Maybe Castle
    -- Church or library but not both
    , cityChurchOrLibrary :: ChurchOrLibrary
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouses :: [House]
    }
```

Reply:
Thanks for the detailed suggestions, feels much easier to comprehend.